src/cockpit/channels/pcp.py

```diff
@@ -0,0 +1,672 @@
+                    except pmapi.pmErr as exc:
+                        logger.debug("Unable to remove profile: instance=%s err=%s", omit_instance, exc)
+
+        factor = 1.0
+        pm_units = pm_desc.units
+        if units:
+            try:
+                [parsed_units, factor] = context.pmParseUnitsStr(units)
+            except pmapi.pmErr as exc:
+                if exc.errno() == c_api.PM_ERR_NAME:
+                    raise ChannelError('not-found', message=f'no such metric: {name}') from None
+                else:
+                    raise ChannelError('internal-error', message=str(exc)) from None
+
+            self.try_convert_unit(context, pm_desc, pm_units)
+
+            if units != parsed_units or factor != 1.0:
+                pm_units = parsed_units
+
+        return MetricInfo(pmid=pm_ids[0],
+                          name=name,
+                          derive=derive,
+                          desc=pm_desc,
+                          factor=factor,
+                          units=pm_units,
+                          instanced=instanced)
+
+    @staticmethod
+    def try_convert_unit(context, pm_desc, pm_units) -> None:
+        """Try to convert a dummy value to validate that the metric is convertible"""
+        dummy = pmapi.pmAtomValue()
+        dummy.d = 0.0
+        try:
+            context.pmConvScale(c_api.PM_TYPE_DOUBLE, dummy, [pm_desc], 0, pm_units)
+        except pmapi.pmErr as exc:
+            raise ChannelError('internal-error', message=str(exc)) from None
+
+    @staticmethod
+    def prepare_archives(archive_dir: str) -> Iterable[ArchiveInfo]:
+        indexes = glob.glob(glob.escape(archive_dir) + '/*.index')
+        for archive_path in indexes:
+            logger.debug('opening archive: %r', archive_path)
+            try:
+                context = pmapi.pmContext(c_api.PM_CONTEXT_ARCHIVE, archive_path)
+                log_label = context.pmGetArchiveLabel()
+                archive_start = float(log_label.start) * 1000
+                yield ArchiveInfo(context, archive_start, archive_path)
+            except pmapi.pmErr as exc:
+                if exc.errno() != c_api.PM_ERR_LOGFILE:
+                    raise ChannelError('not-found', message=f'could not read archive {archive_path}') from None
+
+    @staticmethod
+    def semantic_val(sem_id: int) -> str:
+        if sem_id == c_api.PM_SEM_COUNTER:
+            return "counter"
+        elif sem_id == c_api.PM_SEM_INSTANT:
+            return "instant"
+        elif sem_id == c_api.PM_SEM_DISCRETE:
+            return "discrete"
+
+        return ""
+
+    def send_meta(self, results: 'pmapi.pmResult', context: 'pmapi.pmContext') -> None:
+        metrics = []
+
+        for metric_desc in self.metric_descriptions:
+            desc: Dict[str, Union[str, List[str]]] = {"name": metric_desc.name}
+
+            if metric_desc.derive:
+                desc['derive'] = metric_desc.derive
+
+            if metric_desc.factor == 1.0:
```

Comment:
Uh, exact float comparison.. Is there a more robust way to say "unset"/"default"? Perhaps `None`?

Reply:
It is really bad practice to compare floats for equality. Many values aren't representable exactly as a double (like "3"), so a float should in general be treated as an interval or approximate value. It may work with "1.0" specifically, but better not rely on that. `float | None` sounds better indeed.

Reply:
Ok, I'll investigate that, honestly I would love to get rid of this all. I have experimented a bit more with the high level API

https://gist.github.com/jelly/6f4d683fcd3f8797945f5f90bf0fe5d6

This could be a follow up.

Reply:
> It is really bad practice to compare floats for equality. Many values aren't representable exactly as a double (like "3"), so a float should in general be treated as an interval or approximate value. It may work with "1.0" specifically, but better not rely on that. `float | None` sounds better indeed.

I'm afraid that won't work, as it fails my test now:

```
AssertionError: assert 'byte*1.0' == 'byte'
```

This is due to `pmParseUnitsStr` which returns the parsed_units and a factor. This factor is assigned to in the MetricDescription so we end up in the else where we multiple with 1.0

So in the end we will just move the factor comparison somewhere else.

Reply:
OK. I reluctantly resolve this. This is a really bad API then..